File: algorithms/algorithms-cplusplus/leetcode/Question_0222_Count_Complete_Tree_Nodes.cc
#include <math.h>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int countNodes(TreeNode* root) {
    int left = leftHeight(root);
    int right = rightHeight(root);
    if (left == right) {
      return pow(2, left) - 1;
    }
    return countNodes(root->left) + countNodes(root->right) + 1;
  }

  int leftHeight(TreeNode* root) {
    int res = 0;
    while (root) {
      res++;
      root = root->left;
    }
    return res;
  }

  int rightHeight(TreeNode* root) {
    int res = 0;
    while (root) {
      res++;
      root = root->right;
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0222_Count_Complete_Tree_Nodes.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0222_Count_Complete_Tree_Nodes {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public int countNodes(TreeNode root) {
        int left = leftHeight(root);
        int right = rightHeight(root);
        if (left == right) {
            return (int) Math.pow(2, left) - 1;
        }
        return countNodes(root.left) + countNodes(root.right) + 1;
    }

    private int leftHeight(TreeNode root) {
        int res = 0;
        while (root != null) {
            res++;
            root = root.left;
        }
        return res;
    }

    private int rightHeight(TreeNode root) {
        int res = 0;
        while (root != null) {
            res++;
            root = root.right;
        }
        return res;
    }
}
